File: frontend/src/services/vaccine.service.js
import apiClient from './apiClient';

export async function getCountries() {
  return apiClient.get('/api/vaccines/countries');
}

export async function getManufacturers() {
  return apiClient.get('/api/vaccines/manufacturers');
}

export async function getDosesRequired() {
  return apiClient.get('/api/vaccines/doses-required');
}

export async function getDiseases() {
  return apiClient.get('/api/vaccines/diseases');
}

export async function getBySlug(slug) {
  return apiClient.get(`/api/vaccines/${slug}`);
}

export async function callCreateVaccine(data) {
  return apiClient.post('/api/vaccines', data);
}

export async function callUpdateVaccine(vaccineId, data) {
  return apiClient.put(`/api/vaccines/${vaccineId}`, data);
}

export async function callFetchVaccine(query) {
  return apiClient.get(`/api/vaccines?${query}`);
}

export async function callDeleteVaccine(id) {
  return apiClient.delete(`/api/vaccines/${id}`);
}

// === APIs cho vaccine recommendation theo đối tượng ===

/**
 * Lấy danh sách vaccine phù hợp với tuổi và giới tính
 * @param {number} ageInMonths - Tuổi tính theo tháng
 * @param {string} gender - Giới tính (MALE, FEMALE, ALL)
 */
export async function getRecommendedVaccines(ageInMonths, gender = 'ALL') {
  return apiClient.get(`/api/vaccines/recommended?ageInMonths=${ageInMonths}&gender=${gender}`);
}

/**
 * Lấy danh sách vaccine thiết yếu theo độ tuổi
 * @param {number} ageInMonths - Tuổi tính theo tháng
 */
export async function getEssentialVaccines(ageInMonths) {
  return apiClient.get(`/api/vaccines/essential?ageInMonths=${ageInMonths}`);
}

/**
 * Lấy vaccine theo nhóm đối tượng
 * @param {string} targetGroup - NEWBORN, INFANT, TODDLER, CHILD, TEEN, ADULT, ELDERLY, PREGNANT, ALL
 */
export async function getVaccinesByTargetGroup(targetGroup) {
  return apiClient.get(`/api/vaccines/target-group/${targetGroup}`);
}

/**
 * Lấy vaccine theo danh mục
 * @param {string} category - BASIC_CHILDHOOD, SCHOOL_AGE, ADULT_ROUTINE, PREGNANCY, ELDERLY_CARE, TRAVEL, COVID19
 */
export async function getVaccinesByCategory(category) {
  return apiClient.get(`/api/vaccines/category/${category}`);
}

/**
 * Lấy vaccine theo loại bệnh
 * @param {string} disease
 */
export async function getVaccinesByDisease(disease) {
  return apiClient.get(`/api/vaccines/disease/${disease}`);
}

/**
 * Lấy danh sách tất cả categories
 */
export async function getVaccineCategories() {
  return apiClient.get('/api/vaccines/categories');
}

/**
 * Lấy danh sách tất cả target groups
 */
export async function getTargetGroups() {
  return apiClient.get('/api/vaccines/target-groups');
}

// === APIs cho vaccine cho phụ nữ mang thai ===

/**
 * Lấy vaccine an toàn cho phụ nữ mang thai
 */
export async function getPregnancySafeVaccines() {
  return apiClient.get('/api/vaccines/pregnancy/safe');
}

/**
 * Lấy vaccine nên tiêm trước khi mang thai
 */
export async function getPrePregnancyVaccines() {
  return apiClient.get('/api/vaccines/pregnancy/pre');
}

/**
 * Lấy vaccine nên tiêm sau sinh
 */
export async function getPostPregnancyVaccines() {
  return apiClient.get('/api/vaccines/pregnancy/post');
}

export const callFetchCountry = getCountries;
export const callGetAllCountries = getCountries;
export const callGetAllManufacturers = getManufacturers;
export const callGetAllDosesRequired = getDosesRequired;
export const callGetAllDiseases = getDiseases;
export const callGetAllTargetGroups = getTargetGroups;
export const callGetBySlug = getBySlug;
export const callGetBySku = getBySlug;
export const callFetchVaccineBySlug = getBySlug;

File: backend/src/main/java/com/dapp/backend/service/VaccineService.java
package com.dapp.backend.service;

import com.dapp.backend.dto.mapper.VaccineMapper;
import com.dapp.backend.dto.request.VaccineRequest;
import com.dapp.backend.dto.response.Pagination;
import com.dapp.backend.dto.response.VaccineResponse;
import com.dapp.backend.exception.AppException;
import com.dapp.backend.model.Vaccine;
import com.dapp.backend.repository.VaccineRepository;
import com.dapp.backend.service.spec.VaccineSpecifications;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class VaccineService {
    private final VaccineRepository vaccineRepository;

    public VaccineService(VaccineRepository vaccineRepository) {
        this.vaccineRepository = vaccineRepository;
    }

    private String generateSlug(String name) {
        if (name == null || name.trim().isEmpty()) {
            return "";
        }

        String slug = name.toLowerCase().trim();

        slug = slug.replaceAll("[àáạảãâầấậẩẫăằắặẳẵ]", "a");
        slug = slug.replaceAll("[èéẹẻẽêềếệểễ]", "e");
        slug = slug.replaceAll("[ìíịỉĩ]", "i");
        slug = slug.replaceAll("[òóọỏõôồốộổỗơờớợởỡ]", "o");
        slug = slug.replaceAll("[ùúụủũưừứựửữ]", "u");
        slug = slug.replaceAll("[ỳýỵỷỹ]", "y");
        slug = slug.replaceAll("đ", "d");

        slug = slug.replaceAll("[^a-z0-9\\s-]", "");
        slug = slug.replaceAll("\\s+", "-");
        slug = slug.replaceAll("-+", "-");

        return slug;
    }

    private String generateUniqueSlug(String baseName, Long excludeId) {
        String baseSlug = generateSlug(baseName);
        String uniqueSlug = baseSlug;
        int counter = 1;

        while (true) {
            final String checkSlug = uniqueSlug;
            var existing = vaccineRepository.findBySlug(checkSlug);

            if (existing.isEmpty()) {
                break;
            }

            if (excludeId != null && existing.get().getId() == excludeId) {
                break;
            }

            uniqueSlug = baseSlug + "-" + counter;
            counter++;
        }

        return uniqueSlug;
    }

    public List<String> getAllCountries() {
        return vaccineRepository.findDistinctCountries();
    }

    public List<String> getAllManufacturers() {
        return vaccineRepository.findDistinctManufacturers();
    }

    public List<Integer> getAllDosesRequired() {
        return vaccineRepository.findDistinctDosesRequired();
    }

    public List<String> getAllDiseases() {
        return vaccineRepository.findDistinctDiseases();
    }

    public Vaccine getVaccineBySku(String slug) throws AppException {
        return vaccineRepository.findBySlug(slug)
                .orElseThrow(() -> new AppException("Vaccine not found"));
    }

    public VaccineResponse getVaccineDetailBySlug(String slug) throws AppException {
        Vaccine v = getVaccineBySku(slug);
        return VaccineMapper.toResponse(v);
    }

    public Pagination getAllVaccines(Specification<Vaccine> specification, Pageable pageable) {

        Specification<Vaccine> finalSpec = specification != null
                ? specification.and(VaccineSpecifications.notDeleted())
                : VaccineSpecifications.notDeleted();

        Page<Vaccine> pageVaccine = vaccineRepository.findAll(finalSpec, pageable);
        Pagination pagination = new Pagination();
        Pagination.Meta meta = new Pagination.Meta();
        meta.setPage(pageable.getPageNumber() + 1);
        meta.setPageSize(pageable.getPageSize());
        meta.setPages(pageVaccine.getTotalPages());
        meta.setTotal(pageVaccine.getTotalElements());
        pagination.setMeta(meta);
        pagination.setResult(pageVaccine.getContent().stream().map(VaccineMapper::toResponse).toList());
        return pagination;
    }

    public VaccineResponse createVaccine(VaccineRequest request) throws AppException {

        if (request.getSlug() == null || request.getSlug().trim().isEmpty()) {
            String slug = generateUniqueSlug(request.getName(), null);
            request.setSlug(slug);
        }

        Vaccine vaccine = VaccineMapper.toEntity(request);
        Vaccine savedVaccine = vaccineRepository.save(vaccine);

        return VaccineMapper.toResponse(savedVaccine);
    }

    public VaccineResponse updateVaccine(Long id, VaccineRequest request) throws AppException {

        Vaccine existingVaccine = vaccineRepository.findById(id)
                .orElseThrow(() -> new AppException("Vaccine not found with id: " + id));

        if (request.getSlug() == null || request.getSlug().trim().isEmpty()) {
            String slug = generateUniqueSlug(request.getName(), id);
            request.setSlug(slug);
        }

        VaccineMapper.updateEntity(existingVaccine, request);
        Vaccine updatedVaccine = vaccineRepository.save(existingVaccine);

        return VaccineMapper.toResponse(updatedVaccine);
    }

    public void deleteVaccine(Long id) throws AppException {
        Vaccine vaccine = vaccineRepository.findById(id)
                .orElseThrow(() -> new AppException("Vaccine not found with id: " + id));

        vaccine.setIsDeleted(true);
        vaccineRepository.save(vaccine);
    }

    public List<Vaccine> getVaccinesByName(String name) {
        return vaccineRepository.findAllByName(name);
    }

    // === Methods mới cho việc gợi ý vaccine theo đối tượng ===

    /**
     * Lấy vaccine gợi ý theo độ tuổi và giới tính
     * 
     * @param ageInMonths Tuổi tính theo tháng
     * @param gender      Giới tính (MALE/FEMALE)
     * @return Danh sách vaccine phù hợp
     */
    public List<VaccineResponse> getRecommendedVaccines(Integer ageInMonths, String gender) {
        String genderValue = (gender != null) ? gender.toUpperCase() : "ALL";
        List<Vaccine> vaccines = vaccineRepository.findRecommendedVaccines(ageInMonths, genderValue);
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine an toàn cho phụ nữ mang thai
     */
    public List<VaccineResponse> getPregnancySafeVaccines() {
        List<Vaccine> vaccines = vaccineRepository.findPregnancySafeVaccines();
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine cần tiêm trước khi mang thai
     */
    public List<VaccineResponse> getPrePregnancyVaccines() {
        List<Vaccine> vaccines = vaccineRepository.findPrePregnancyVaccines();
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine cần tiêm sau sinh
     */
    public List<VaccineResponse> getPostPregnancyVaccines() {
        List<Vaccine> vaccines = vaccineRepository.findPostPregnancyVaccines();
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine theo nhóm đối tượng
     * 
     * @param targetGroup NEWBORN, INFANT, TODDLER, CHILD, TEEN, ADULT, ELDERLY
     */
    public List<VaccineResponse> getVaccinesByTargetGroup(String targetGroup) {
        List<Vaccine> vaccines = vaccineRepository.findByTargetGroup(targetGroup.toUpperCase());
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine theo danh mục
     * 
     * @param category BASIC_CHILDHOOD, SCHOOL_AGE, ADULT_ROUTINE, PREGNANCY,
     *                 ELDERLY_CARE, TRAVEL, COVID19
     */
    public List<VaccineResponse> getVaccinesByCategory(String category) {
        List<Vaccine> vaccines = vaccineRepository.findByCategory(category.toUpperCase());
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine theo loại bệnh
     */
    public List<VaccineResponse> getVaccinesByDisease(String disease) {
        List<Vaccine> vaccines = vaccineRepository.findByDisease(disease);
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy vaccine thiết yếu theo độ tuổi
     */
    public List<VaccineResponse> getEssentialVaccinesByAge(Integer ageInMonths) {
        List<Vaccine> vaccines = vaccineRepository.findEssentialVaccinesByAge(ageInMonths);
        return vaccines.stream().map(VaccineMapper::toResponse).toList();
    }

    /**
     * Lấy danh sách categories
     */
    public List<String> getAllCategories() {
        return vaccineRepository.findDistinctCategories();
    }

    /**
     * Lấy danh sách target groups
     */
    public List<String> getAllTargetGroups() {
        return vaccineRepository.findDistinctTargetGroups();
    }
}
